proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,582 @@
+enabled and all the known `Subscription`s are in the same process, then the message is only published
+intra-process. This remains identical to the current implementation.
+
+### Creating a publisher
+
+1. User calls `Node::create_publisher<MessageT>(...)`.
+2. This boils down to `NodeTopics::create_publisher(...)`, where a `Publisher` is created through the factory.
+3. Here, if intra-process communication is enabled, eventual intra-process related variables are initialized through the
+   `Publisher::SetupIntraProcess(...)` method.
+4. Then the `IntraProcessManager` is notified about the existence of the new `Publisher` through the method
+   `IntraProcessManager::add_publisher(PublisherBase::SharedPtr publisher, PublisherOptions options)`.
+5. `IntraProcessManager::add_publisher(...)` stores the `Publisher` information in an internal structure of
+   type `PublisherInfo`. The structure contains information about the `Publisher`, such as its QoS and its topic name, and
+   a weak pointer for the `Publisher` object.
+   An `uint64_t pub_id` unique within the `rclcpp::context` is assigned to the `Publisher`.
+   The `IntraProcessManager` contains a `std::map<uint64_t, PublisherInfo>` object where
+   it is possible to retrieve the `PublisherInfo` of a specific `Publisher` given its id.
+   The function returns the `pub_id`, that is stored within the `Publisher`.
+
+If the `Publisher` QoS is set to `transient local`, then the `Publisher::SetupIntraProcess(...)` method will also create
+a ring buffer of the size specified by the depth from the QoS.
+
+### Creating a subscription
+
+1. User calls `Node::create_subscription<MessageT>(...)`.
+2. This boils down to `NodeTopics::create_subscription(...)`, where a `Subscription` is created through the factory.
+3. Here, if intra-process communication is enabled, intra-process related variables are initialized through the
+   `Subscription::SetupIntraProcess(...)` method. The most relevant ones being the ring buffer and the waitable object.
+4. Then the `IntraProcessManager` is notified about the existence of the new `Subscription` through the method
+   `IntraProcessManager::add_subscription(SubscriptionBase::SharedPtr subscription, SubscriptionOptions options)`.
+5. `IntraProcessManager::add_subscription(...)` stores the `Subscription` information in an internal structure
+   of type `SubscriptionInfo`. The structure contains information about the `Subscription`, such as its QoS, its topic name
+   and the type of its callback, and a weak pointer for the `Subscription` object.
+   An `uint64_t sub_id` unique within the `rclcpp::context` is assigned to the `Subscription`.
+   The `IntraProcessManager` contains a `std::map<uint64_t, SubscriptionInfo>` object where
+   it is possible to retrieve the `SubscriptionInfo` of a specific `Subscription` given its id.
+   There is also an additional structure `std::map<uint64_t, std::pair<std::set<uint64_t>, std::set<uint64_t>>>`.
```

Comment:
This structure is a substitute for the `IDTopicMap subscription_ids_by_topic_` that is present in the current implementation.

Since we want to support more QoS configurations it is not sufficient anymore to select the nodes where to push the messages looking only at the topic name